Sort files into alphabetical order before loading them

AaronC81 opened the pull request:
First off, thanks a bunch for this plugin - much more convenient than writing a full plugin of my own for making small improvements or extensions to Obsidian!

Currently, the two devices which I use Obsidian on load my JS files in a different order. Some of my scripts depend on others being loaded first, so this caused some problems. 

As an example, I created three new JS files, `a.js`, `b.js` and `c.js`:

- Linux laptop: c, b, a
- Android phone: c, a, b

This PR sorts the list of loaded files alphabetically by their full paths, so they load in a consistent order across all platforms. I've tested this on my laptop and phone and this change seems to work OK on both :)

SamLewis0602 commented:
Hey! Thanks for contributing!

I added a commit to mention this in the docs and then realized that I'm a bit confused why the order would matter. I see your mention of dependencies loading in different orders, but the way CustomJS works is it builds an object with all of your classes and then assigns the object to the global window. So all of your classes effectively become available at the same time. Could you expand a bit on how your scripts depend on each other? I just want to fully understand the use case. Thanks!

AaronC81 commented:
Sure! The setup I have follows roughly this structure:

sidebar_extensions.js:

```js
function create_sidebar_extensions() {
  class SidebarExtensions {
    register(id) {
      // ...
    }
  } 

  window.sidebar_extensions = new SidebarExtensions();
}
```

tasks_extension.js:

```js
function register_task_extension() {
  const ext = window.sidebar_extensions.register("task");
  // ...
}
```

This worked fine on desktop, but on my phone the sidebar extension was never registered. To try to solve this, I added a busy-wait for `window.sidebar_extensions` to exist, hoping that the scripts would run in parallel:

```js
function register_task_extension() {
  while (!window.sidebar_extensions);
  const ext = window.sidebar_extensions.register("task");
  // ...
}
```

The desktop app still worked but the mobile app locked up completely, so I could deduce from this that tasks_extension.js was running after sidebar_extensions.js on desktop, but before on mobile. After that I tried the a, b, c experiment which I described in the initial comment, verifying that they run in a different order, and I can only assume that was what was causing this issue.

With this patch, I renamed sidebar_extensions.js to aaaa_sidebar_extensions.js and it now works on both platforms, without the busy-wait.

I hope that helps :)

nkoder commented:
As this solution looks very unusual to me, I wonder if your problem @AaronC81 can be solved by deferring access to extensions to the moment after all scripts are loaded 🤔 

I have too few info about your setup and needs to give a specific example, so let's have this abstract one:

Let's imagine two scripts, `script_a.js` and `script_b.js`, each requiring the other one to work. Maybe you can use it like this:

```js
// in script_a.js
function do_something_in_script_A() {
    return {
        initA: function () {
           const somethingFromB = window.script_b.doSomething();
           // …
        },
        doStuff: function () {
            // …
        }
    };
}
```

and 

```js
// in script_b.js
function do_something_in_script_B() {
    return {
        initB: function () {
           const somethingFromA = window.script_a. doStuff();
           // …
        },
        doSomething: function () {
            // …
        }
    };
}
```

and then in every `dataviewjs` code blck for example:

```js
// …
scriptA.initA();
scriptB.initB();
// …
```

I am aware my example is not the best, written by hand without any code to check it out. My point is that maybe you can **not** perform any actions when script is evaluated/loaded, but instead expose a function which will allow you to perform those actions e.g. in `dataviewjs` block where you know all dependencies are loaded already 🤔

AaronC81 commented:
@nkoder Thanks for your reply, that would be a great solution for my use-case in a dataview. However, I am not using CustomJS for extending dataviews; instead, I'm using it to create my own mini plugins which actually extend Obsidian's UI.

The aaaa_sidebar_extensions.js script creates `window.sidebar` to provide the framework for doing this, and then my other scripts can register their sidebar extensions using that framework by calling methods on `window.sidebar`:

![image](https://user-images.githubusercontent.com/3321773/134819948-f44774f5-199c-4b6c-aaaa-7e52cb37a3ef.png)

The "initialize when everything is done" approach which you've suggested may be a good plan, if I can find some event which fires when Obsidian has completely finished loading and other scripts have run. Thank you :)

If it helps, here is the entire set of scripts I am using to achieve this: https://gist.github.com/AaronC81/7cac786aa9a263e6d1bebe75494456c1

(N.B. If what I'm doing here is too far from what CustomJS supports, then I entirely understand not merging this!)

SamLewis0602 commented:
Thanks for explaining! So the sticking point is that you're adding to window and then referencing that global in the other script. It's a bit hacky but so is CustomJS itself 😁 and loading scripts alphabetically doesn't hurt anything so I don't mind merging it.

However, I think I'd like to make a change, which is to gather all scripts from both settings options and then sort them by filename. Rather than sorting individual scripts by filename and the scripts directory by path. That would make the behavior more consistent. I'm gonna add that in real quick

SamLewis0602 commented:
Okay just pushed up that change. I'm a bit skeptical where the `.last()` is coming from since that's not standard js, Obisidian must have extended the prototype Array object to add that. Will you test the latest version for me to double check everything works?

Here's an example of the sorting before and after:
<img width="291" alt="image" src="https://user-images.githubusercontent.com/15079088/134906794-1480955b-b01c-4683-85f4-f982e0c6ccc3.png">

SamLewis0602 commented:
Yep, turns out they did add the `Array.last()` method

https://github.com/obsidianmd/obsidian-api/blob/5af3ef982328a2dd0a2f5242f3c74e2f45e07896/obsidian.d.ts#L17

AaronC81 commented:
Super, thanks - still works on my laptop and phone!

SamLewis0602 commented:
Merged. I'll get a release out later today. Thanks again for contributing!
